### DISCUSSION TITLE
Abstract file system - readFile not called for layouts and partials

### DISCUSSION THREAD

**Discussion by TimGeudens** [harttle/liquidjs#424] [Q&A] [answered]
Hi,

We are creating a POC with liquidjs to see if it's compatible with our needs. 
One of the requirements is to load the templates from an external database. The abstract file system looks to be perfect for that. 
I'm trying to use it like this: 
`const engine = new Liquid({
  extname: '.liquid',
  globals: { title: 'LiquidJS POC' },
  fs: {
    readFileSync() {
      throw new Error('should not be used');
    },
    async readFile(file) {
      if (templates[file]) {
        return templates[file];
      }
      const result = await getTemplate(file.indexOf('.liquid') > -1 ? file : `${file}.liquid`);
      templates[file] = result[0].text || '';
      return result[0].text || '';
    },
    existsSync() {
      return true;
    },
    exists() {
      return true;
    },
    resolve(root, file) {
      return file;
    },
  },
});`

getTemplate fetches the liquidjs text from a DB.
This works perfect for our templates file. However when that template file includes a layout or partial file, I get the error:
`ENOENT: Failed to lookup "html.liquid" in ".", file:case`

It's not calling the readFile function, but we also want to get the layout and partial files from the DB. 
I have found that the 4th parameter of the this.candidates function is the "cause" of the error: https://github.com/harttle/liquidjs/blob/d4df3a0b5365d57c0cb5ddb8fe034e2a4f4946c0/src/fs/loader.ts#L32
When setting it to always be false I can get it to work. 

Is there something I'm doing wrong, maybe some parameter that is missing? 
I saw this recent post (https://github.com/harttle/liquidjs/discussions/422) but couldn't figure out if it has anything to do with my issue. 

Regards

**Comment by harttle** [comment, marked as answer]
Sorry recently master is broken but now it's fixed on v9.28.3, try that again and please let me know if it's still not working. And feel free to open issues when you get other unexpected errors.

**Reply by TimGeudens** [reply]
That version fixes the problem, thank you for the quick reply!
